crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -13,14 +20,72 @@ pub enum Trade {
 pub struct Fulfillment {
     pub order: competition::Order,
     /// The amount executed by this fulfillment. See
-    /// [`competition::order::Partial`]. If the order is not partial, the target
-    /// amount must equal the amount from the order.
+    /// [`competition::order::Partial`]. If the order is not partial, the
+    /// executed amount must equal the amount from the order.
     pub executed: competition::order::TargetAmount,
 }
 
+impl Fulfillment {
+    /// Calculate the settlement contract input and output amounts executed by
+    /// this trade.
+    pub fn execution(&self, clearing_prices: &ClearingPrices) -> Result<Execution, Error> {
+        let input = self.executed.to_asset(&self.order);
+        let (input_price, output_price) = {
+            let sell_price = clearing_prices
+                .0
+                .get(&self.order.sell.token)
+                .ok_or(Error::ClearingPriceMissing)?
+                .to_owned();
+            let buy_price = clearing_prices
+                .0
+                .get(&self.order.buy.token)
+                .ok_or(Error::ClearingPriceMissing)?
+                .to_owned();
+            match self.order.side {
+                order::Side::Buy => (buy_price, sell_price),
+                order::Side::Sell => (sell_price, buy_price),
+            }
+        };
+        let output = eth::Asset {
+            amount: match self.order.kind {
+                order::Kind::Market => input
+                    .amount
+                    .checked_mul(input_price)
+                    .ok_or(Error::Overflow)?
+                    .checked_ceil_div(&output_price)
+                    .ok_or(Error::Overflow)?,
+                order::Kind::Limit { .. } => todo!(),
+                order::Kind::Liquidity => todo!(),
+            },
+            token: match self.order.side {
+                order::Side::Buy => self.order.sell.token,
+                order::Side::Sell => self.order.buy.token,
+            },
```

Comment:
The sell token is always the token that flow into the contract and the buy token always flows out. The order `side` is only needed to know which amount needs to be computed with the price.
A `buy` order has a fixed buy amount so we have to use the clearing price to compute the sell amount. And a `sell` order has a fixed sell amount and we need to compute the buy amount from it.